lib/irc/ClientPool.js

```diff
@@ -276,26 +284,28 @@ ClientPool.prototype._onClientDisconnected = function(bridgedClient) {
     // remove ref to the disconnected client so it can be GC'd. If we don't do this,
     // the timeout below holds it in a closure, preventing it from being GC'd.
     bridgedClient = undefined;
-    setTimeout(function() {
-        cli.connect().then(function() {
-            log.info(
-                "<%s> Reconnected %s@%s", cli._id, cli.nick, cli.server.domain
-            );
-            if (chanList.length > 0) {
-                log.info("<%s> Rejoining %s channels", cli._id, chanList.length);
-                chanList.forEach(function(c) {
-                    cli.joinChannel(c);
-                });
-            }
-            self._sendConnectionMetric(cli.server);
-        }, function(e) {
-            log.error(
-                "<%s> Failed to reconnect %s@%s", cli._id, cli.nick, cli.server.domain
-            );
-        });
-    }, RECONNECT_TIME_MS);
+    this._reconnectQueue.enqueue(cli._id, {cli, chanList});
 };
 
+ClientPool.prototype._reconnectClient = function({cli, chanList}) {
+    return cli.connect().then(() => {
+        log.info(
+            "<%s> Reconnected %s@%s", cli._id, cli.nick, cli.server.domain
+        );
+        if (chanList.length > 0) {
```

Comment:
If the `chanList` is empty, perhaps we could bail out before even adding this to the QueuePool?

Reply:
Good idea :thinking: